### PR TITLE
[RISCV] Use software guarded branch for indirect jump table branch.

### PR DESCRIPTION
**Pull request by yetingk**
When Zicfilp enabled, indirect jump table branch should be a software guarded branch.

### CODE REVIEW

**Comment by preames** [comment]
I think I'm missing something fundamental about the implementation status here.  The GPRJALR register class used by BRIND (the non-guarded form) doesn't seem to exclude X7.  Given that, how is it useful to define a guarded form which requires X7?

Has there been an assembly syntax defined for the SW guarded branch?  An alias or anything else?  I don't see one in the linked specification (despite defining an alias for landing pad).  Having disassembler support for guarded branches would normally be a first step...

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td [hunk 1]
```diff
@@ -46,6 +46,7 @@ def SDT_RISCVIntBinOpW : SDTypeProfile<1, 2, [
 def SDT_RISCVIntShiftDOpW : SDTypeProfile<1, 3, [
   SDTCisSameAs<0, 1>, SDTCisSameAs<0, 2>, SDTCisVT<0, i64>, SDTCisVT<3, i64>
 ]>;
+def SDT_RISCVSWGuardedBrind : SDTypeProfile<0, -1, [SDTCisVT<0, iPTR>]>;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.td:49]: Why is this different than

```
def SDTBrind : SDTypeProfile<0, 1, [        // brind                             
  SDTCisPtrTy<0>                                                                 
]>;
```

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td [hunk 2]
```diff
@@ -1916,6 +1926,12 @@ def : Pat<(binop_allwusers<add> GPR:$rs1, (AddiPair:$rs2)),
                  (AddiPairImmSmall AddiPair:$rs2))>;
 }
 
+let Predicates = [HasStdExtZicfilp] in {
+def : Pat<(riscv_sw_guarded_brind GPRX7:$rs1), (PseudoBRINDX7 GPRX7:$rs1, 0)>;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.td:1930]: Is this the same as the code on line 1474?

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**